File: example/android/app/src/main/kotlin/dev/brewkits/native_workmanager_example/workers/ImageCompressWorker.kt
package dev.brewkits.native_workmanager_example.workers

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import dev.brewkits.kmpworkmanager.background.domain.AndroidWorker
import dev.brewkits.kmpworkmanager.background.domain.WorkerResult // Added import for WorkerResult
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import java.io.File
import java.io.FileOutputStream

class ImageCompressWorker : AndroidWorker {
    override suspend fun doWork(input: String?): WorkerResult { // Changed return type to WorkerResult
        try {
            // Parse JSON input
            val json = Json.parseToJsonElement(input ?: "{}")
            val config = json.jsonObject

            val inputPath = config["inputPath"]?.jsonPrimitive?.content
                ?: return WorkerResult.Failure("Input path or output path missing")
            val outputPath = config["outputPath"]?.jsonPrimitive?.content
                ?: return WorkerResult.Failure("Input path or output path missing")
            val quality = config["quality"]?.jsonPrimitive?.content?.toIntOrNull()
                ?: 85

            // Load image
            val bitmap = BitmapFactory.decodeFile(inputPath)
                ?: return WorkerResult.Failure("Failed to load image from input path")

            // Compress and save
            val outputFile = File(outputPath)
            outputFile.parentFile?.mkdirs()

            FileOutputStream(outputFile).use { out ->
                bitmap.compress(Bitmap.CompressFormat.JPEG, quality, out)
            }

            bitmap.recycle()
            return WorkerResult.Success()

        } catch (e: Exception) {
            println("ImageCompressWorker error: ${e.message}")
            return WorkerResult.Failure("Image compression failed: ${e.message}")
        }
    }
}

File: example/ios/Runner/ImageCompressWorker.swift
import Foundation
import UIKit

/// Errors thrown by ImageCompressWorker.
enum ImageCompressError: Error {
    case invalidInput(String)
    case processingFailed(String)
}

/// Demo custom native worker that compresses a JPEG image.
///
/// Shows how to register a custom worker with IosWorkerFactory from AppDelegate.swift.
///
/// Expected JSON input:
/// ```json
/// {
///   "inputPath": "/path/to/input.jpg",
///   "outputPath": "/path/to/output.jpg",
///   "quality": 0.85
/// }
/// ```
class ImageCompressWorker: IosWorker {
    func doWork(input: String?, env: WorkerEnvironment) async throws -> WorkerResult {
        // Parse JSON input
        guard let inputString = input,
              let data = inputString.data(using: .utf8),
              let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
              let inputPath = json["inputPath"] as? String,
              let outputPath = json["outputPath"] as? String else {
            throw ImageCompressError.invalidInput("Missing required input parameters: inputPath, outputPath")
        }

        let quality = (json["quality"] as? Double) ?? 0.85

        // Load image
        guard let image = UIImage(contentsOfFile: inputPath) else {
            throw ImageCompressError.processingFailed("Failed to load image from \(inputPath)")
        }

        // Compress
        guard let compressedData = image.jpegData(compressionQuality: quality) else {
            throw ImageCompressError.processingFailed("Failed to compress image")
        }

        // Save
        let outputURL = URL(fileURLWithPath: outputPath)
        do {
            try FileManager.default.createDirectory(
                at: outputURL.deletingLastPathComponent(),
                withIntermediateDirectories: true
            )
            try compressedData.write(to: outputURL)
        } catch {
            throw ImageCompressError.processingFailed("Failed to save compressed image: \(error.localizedDescription)")
        }

        let originalSize = (try? FileManager.default.attributesOfItem(atPath: inputPath)[.size] as? Int) ?? 0
        let compressedSize = compressedData.count

        return WorkerResult.success(
            message: "Compressed image saved to \(outputPath)",
            data: [
                "outputPath": outputPath,
                "originalSize": originalSize,
                "compressedSize": compressedSize,
            ]
        )
    }
}
